### ISSUE TITLE
[Due for payment 2025-02-19] Prevent Self-Approval: Fix Unexpected Behavior for Historical Workspaces and Update Existing Approvers

### ISSUE THREAD

**Issue by marcaaron** [Expensify/App#53799]
cc @JmillsExpensify coming from our DM

While we have the ability to set "Prevent self approval" on a workspace. There are some ways in which this feature works unexpectedly in NewDot. Specifically, the following scenario is possible and should not be:

## Historical workspace behavior

**Problem**

- User has historical workspaces with themselves already set as the approver. 
- Self approval is then enabled. 
- What should we do after that? They should not be able to approve their own expenses, but we need some way to "fix" the reports currently submitted to themselves.

**Solution**

- Create a modal that warns the user that any users who are currently approving their own reports will be removed as the manager.
- Set the approver back to the "default approver" as shown here:

![image](https://github.com/user-attachments/assets/6ab6cd5a-476a-4204-9d77-38280270fdfb)

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

**Comment by marcaaron** [comment]
Specifically we need design help on:

>Create a modal that warns the user that any users who are currently approving their own reports will be removed as the manager.

**Comment by marcaaron** [comment]
QA steps would be something like:

1. As User A enable Rules on a workspace
2. Invite User B
3. In Workflows create a rule for User B that has them set as their own approver
![Screenshot 2024-12-11 at 12 34 40 PM](https://github.com/user-attachments/assets/5481259e-d041-4b46-9651-a47c0ded5283)
5. In Rules select the Prevent Self Approvals option
![Screenshot 2024-12-11 at 12 34 54 PM](https://github.com/user-attachments/assets/169a4a88-3963-4b32-a548-86ff49811310)
6. Modal is triggered since now User B can't be their own approver. We will be setting their approver to the default approver if the modal is "confirmed".

**Comment by dannymcclain** [comment]
Perfect! Thank you.

**Comment by dannymcclain** [comment]
Ok so I think we have two options here—we can go classic confirmation dialog or use one of our modals with the illustration.

### Confirmation dialog
![image](https://github.com/user-attachments/assets/dd187823-41a4-435c-aaab-bb14c83be721)

### Illustrated modal
![image](https://github.com/user-attachments/assets/1283b1e6-0fac-4a22-8675-5aef3ee85738)

@Expensify/design any preference here? I feel like the regular confirmation dialog feels most appropriate for this, right?

> [!NOTE]
> Need help with the copy here

**Comment by shawnborton** [comment]
Yeah, I tend to agree - just a simple, matter-of-fact dialog seems appropriate.

**Comment by dannymcclain** [comment]
And actually, I just realized I did this one wrong. I think it should actually be like this:
![image](https://github.com/user-attachments/assets/36a9f26d-8628-455d-9b3e-f8a36cd5dbee)

Where we give you the warning and the choice. @shawnborton @dubielzyk-expensify, the question is, should this button be `danger`?

**Comment by dannymcclain** [comment]
Yeah I feel the same way. Let's make @dubielzyk-expensify weigh in! `:gavel:`

**Comment by dubielzyk-expensify** [comment]
Yeah, I'm with ya. Don't feel strongly. Danger button seems perhaps a bit too harsh for what it is. cc @jamesdeanexpensify for copy!?

**Comment by dannymcclain** [comment]
I think we can go with the green button for this one—even if some workflows are reset to the default approver, no information or anything is actually being destroyed. Nothing is going to stop working. So I agree with Jon the the danger button feels too harsh.

**Comment by jamesdeanexpensify** [comment]
Slight tweaks!

> **Prevent self-approvals?**
> Any members currently approving their own expenses will be replaced with the default approver for this workspace (email/phone).

**Comment by dannymcclain** [comment]
Update mocks based on new copy from James!

![image](https://github.com/user-attachments/assets/66a9dc79-297a-49ef-8082-449fa8566d09)

**Comment by dubielzyk-expensify** [comment]
Thank McClain!

**Comment by trjExpensify** [comment]
👋 I think I have a few questions about this. But maybe to start, it would be helpful if someone could list out the numbered steps of the proposed solution in full? Are the QA steps [here](https://github.com/Expensify/App/issues/53799#issuecomment-2537342411) the extent of the proposal, or is there more to it? 

Thanks!

**Comment by dannymcclain** [comment]
Any update on this one? Do we have clear next steps to move this along?

**Comment by JmillsExpensify** [comment]
Btw, we need to find someone to pick this now that Marc is on leave.

**Comment by JmillsExpensify** [comment]
Still need to find a volunteer. This is being highlighted as a hot pick in the weekly update.

**Comment by lakchote** [comment]
I have a few questions regarding this:

What happens to open or submitted reports if the user was self-approving them?
- does toggling Prevent Self Approvals instantly re-route them to the default approver?

What if no valid manager can be set (e.g., the user is alone in the workspace/no other manager exists OR the user is the default workspace owner)?
- should we block enabling Prevent Self Approvals entirely?

cc @JmillsExpensify @trjExpensify

**Comment by JmillsExpensify** [comment]
> does toggling Prevent Self Approvals instantly re-route them to the default approver?

Good question. I think the "ideal" is that we'd re-route them, though that's also not very consistent with how we handle changes platform-wide. So as a result, I think the most consistent thing is to prevent self-approvals moving forward. 

> should we block enabling Prevent Self Approvals entirely?

Yes, I think we should. We're similarly locking features on this same page when someone tries to enable something that isn't possible (or that will error out).

**Comment by lakchote** [comment]
>> should we block enabling Prevent Self Approvals entirely?
>
>Yes, I think we should. We're similarly locking features on this same page when someone tries to enable something that isn't possible (or that will error out).
>

@JmillsExpensify the default workspace workflow is the one where everyone submit to the workspace owner/approver.

I've tried locking the setting if the user that wanted to prevent self approvals, was the workspace owner himself.
However I decided to not purse that option. Why? Because it'd mean everyone would have a specific additional workflow where the workspace owner submits to another admin. Let me know if though that's what we want! I'll adjust accordingly.

**Comment by JmillsExpensify** [comment]
Took the discussion to Slack quickly to make sure we're on the same page.

**Comment by JmillsExpensify** [comment]
@lakchote I think we resolved this discussion/question in Slack, is that right?

**Comment by lakchote** [comment]
> [@lakchote](https://github.com/lakchote) I think we resolved this discussion/question in Slack, is that right?
>

Yes, we resolved that discussion in Slack.
However, @trjExpensify suggested [here](https://github.com/Expensify/App/pull/55740#discussion_r1935793041) that we do not lock the option as we don't currently do it in OldDot. Should we get rid of this or not?

**Comment by JmillsExpensify** [comment]
I feel like locking is a front-end paradigm, and we've experienced an issue in practice where locking would have prevented this issue and prevented a support volume. So I'm for moving forward with it. 

> Currently, another problem we have is that the next steps are not calculated properly on NewDot, when prevent self approvals is enabled. On the self approver side, it's computed correctly. However, on the policy's owner side, it says we're waiting on the previous self approver to approve.

My opinion on this one is that it's a separate problem and we are working on solving it as part of the next steps in Auth project.

**Comment by trjExpensify** [comment]
>I feel like locking is a front-end paradigm, and we've experienced an issue in practice where locking would have prevented this issue and prevented a support volume. So I'm for moving forward with it.

Should we prevent it from being enabled on OldDot as well then, so the two settings are aligned? Otherwise you can enable this on OldDot and it'll enable it on NewDot when it should otherwise be "locked" with one member? 🤔

**Comment by lakchote** [comment]
> > I feel like locking is a front-end paradigm, and we've experienced an issue in practice where locking would have prevented this issue and prevented a support volume. So I'm for moving forward with it.
> 
> Should we prevent it from being enabled on OldDot as well then, so the two settings are aligned? Otherwise you can enable this on OldDot and it'll enable it on NewDot when it should otherwise be "locked" with one member? 🤔

I can raise another PR to do this, and make things consistent.

**Comment by trjExpensify** [comment]
Sounds good!

**Comment by lakchote** [comment]
Web-E [PR](https://github.com/Expensify/Web-Expensify/pull/45358) logic has been merged.

Asked for C+ review again in https://expensify.slack.com/archives/C02NK2DQWUX/p1738693963005669?thread_ts=1738144137.928959&cid=C02NK2DQWUX.

**Comment by lakchote** [comment]
Waiting for an internal engineer [review](https://github.com/Expensify/App/pull/55740#issuecomment-2637410179), and it should be merged!

**Comment by garrettmknight** [comment]
Taking over for payment

**Comment by alitoshmatov** [comment]
@garrettmknight Sorry, I forgot to comment here. I was C+ reviewer in the [PR](https://github.com/Expensify/App/pull/55740#issuecomment-2652454679). Can we handle payment for me?

**Comment by garrettmknight** [comment]
**Payment Summary:**

- @alitoshmatov owed $250 via NewDot

**Comment by JmillsExpensify** [comment]
$250 approved for @alitoshmatov
